[Pages] Update Next.js guide for `@cloudflare/next-on-pages` v1 (#8939)
cloudflare/cloudflare-docs: content/pages/framework-guides/deploy-a-nextjs-site.md

# Deploy a Next.js site

> [Next.js](https://nextjs.org) is an open-source React framework for creating websites and apps. In this guide, you will create a new Next.js application and deploy it using Cloudflare Pages.

Reviewer: ```suggestion
[Next.js](https://nextjs.org) is an open-source React framework for creating websites and apps. In this guide, you will create a new Next.js application and deploy it using Cloudflare Pages.

This guide will instruct you how to deploy:

* Full-stack Next.js projects which use the [Edge Runtime](https://nextjs.org/docs/app/api-reference/edge). 
* Next.js projects which can be [statically exported](https://nextjs.org/docs/app/building-your-application/deploying/static-exports).
```

## Consider if you need the Edge Runtime

> Cloudflare Pages supports full-stack Next.js projects which use the [Edge Runtime](https://nextjs.org/docs/app/api-reference/edge), or any projects which can be [statically exported](https://nextjs.org/docs/app/building-your-application/deploying/static-exports). The Edge Runtime allows applications to use server-side features such as [Edge API Routes](https://nextjs.org/docs/api-routes/edge-api-routes), server-side rendering (SSR) pages with [`getServerSideProps()`](https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props), [Server Components](https://nextjs.org/docs/getting-started/react-essentials#server-components), and [Middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware).

Reviewer: ```suggestion
The Edge Runtime allows applications to use server-side features such as [Edge API Routes](https://nextjs.org/docs/api-routes/edge-api-routes), server-side rendering (SSR) pages with [`getServerSideProps()`](https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props), [Server Components](https://nextjs.org/docs/getting-started/react-essentials#server-components), and [Middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware).
```

For more information about the Edge Runtime, refer to [the official Next.js documentation](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes) which explains the differences between the Edge Runtime and traditional Node.js servers, or [read the Cloudflare announcement blog post](https://blog.cloudflare.com/next-on-pages).

{{<render file="_tutorials-before-you-start.md">}}

> ## Using the Edge Runtime

Reviewer: ```suggestion
## Use the Edge Runtime
```

> ### 1. Select your Next.js app

Reviewer: ```suggestion
### 1. Select your Next.js project
```

> If you already have a Next.js project that you wish to deploy, change to its directory and proceed to the next step. Otherwise, you can use `create-next-app` to start a new one.

Reviewer: ```suggestion
If you already have a Next.js project that you wish to deploy, change to its directory and proceed to the next step. Otherwise, use `create-next-app` to create a new Next.js project:
```

```sh
$ npx create-next-app my-app
```

> After creating your project, a new `my-app` directory will be generated using the official default template. Switch to this directory to continue.

Reviewer: ```suggestion
After creating your project, a new `my-app` directory will be generated using the official default template. Change to this directory to continue.
```

```sh
$ cd my-app
```

### 2. Configure the application to use the Edge Runtime

The default template uses traditional Node.js-powered routes that are not supported on Cloudflare Pages. To run your application, you need to opt into the Edge Runtime for any routes that have server-side functionality (e.g. API routes or pages that use `getServerSideProps`). To do this, you need to export a `runtime` [route segment config](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#runtime) option from each route's file.

```js

> export const runtime = "edge";

Author: ```suggestion
export const runtime = 'edge';
```

```ts
---
filename: app/api/hello/route.ts
---
import { cookies } from 'next/headers';

export const runtime = 'edge';

export async function GET(request: Request) {
	const cookieStore = cookies();
	const token = cookieStore.get('token');

	return new Response('Hello, Next.js!', {
		status: 200,
		headers: { 'Set-Cookie': `token=${token}` },
	});
}
```

{{</tab>}}
{{</tabs>}}

For more examples of this and for Next.js versions prior to v13.3.1, take a look at the `@cloudflare/next-on-pages` [examples document](https://github.com/cloudflare/next-on-pages/blob/main/docs/examples.md). Additionally, ensure that your application is not using any [unsupported APIs](https://nextjs.org/docs/app/api-reference/edge#unsupported-apis) or [features](https://github.com/cloudflare/next-on-pages/blob/main/docs/supported.md).

{{<render file="_create-github-repository_no_init.md">}}

### 3. Deploy your application to Cloudflare Pages

To deploy your application to Cloudflare Pages, you need to install the `@cloudflare/next-on-pages` package. This library builds your Next.js project in a format that can be deployed to Pages, and handles the runtime logic for your application.

```sh

> $ npm install @cloudflare/next-on-pages --save-dev

Reviewer 2: For consistency with the other guides (and as a personal preference) 🙂 

```suggestion
$ npm install --save-dev @cloudflare/next-on-pages
```

   {{<table-wrap>}}

   | Configuration option | Value                              |
   | -------------------- | ---------------------------------- |
   | Production branch    | `main`                             |

>    | Build command        | `npx @cloudflare/next-on-pages@v1` |

Author: ```suggestion
   | Build command        | `npx @cloudflare/next-on-pages@1` |
```

   | Build directory      | `.vercel/output/static`            |

   {{</table-wrap>}}

4. Next.js requires Node.js v16 or later to build successfully. To set your Node version, go to **Settings** in your Pages project > **Environment Variables (advanced)** section and add a `NODE_VERSION` variable with a value of `16` or greater.

> 5. Click on **Save and Deploy** to start the deployment (this first deployment won't be fully functional as the next step is also necessary).

Reviewer: ```suggestion
5. Click on **Save and Deploy** to start the deployment. This first deployment will not be fully functional as the next step is also necessary.
```

> 6. Go to the **Pages project settings** page (_Settings_ > _Functions_ > _Compatibility Flags_), **add the `nodejs_compat` flag** for both production and preview, and make sure that the **Compatibility Date** for both production and preview is set to at least `2022-11-30`.

Reviewer: ```suggestion
6. In your Pages project, go to **Settings** > **Functions** > **Compatibility Flags**.
7. Configure a `nodejs_compat` flag for both production and preview .
8. Above **Compability Flags**, go to **Compatibility Date**  and configure a compatibility date that is at least `2022-11-30` for both production and preview.
```

{{<Aside type="note" header="Note">}}

> The `@cloudflare/next-on-pages` CLI transforms the Edge Runtime components of your project into a `_worker.js` file which is deployed with [Pages Functions](/pages/platform/functions/advanced-mode/). The library attempts to deduplicate code that would otherwise result in your project quickly hitting the [script size limit](/workers/platform/limits/#worker-size). If you notice any bugs, please let us know by filing a [GitHub issue](https://github.com/cloudflare/next-on-pages/issues/).

Reviewer: ```suggestion
The `@cloudflare/next-on-pages` CLI transforms the Edge Runtime components of your project into a `_worker.js` file which is deployed with [Pages Functions](/pages/platform/functions/advanced-mode/). The library attempts to deduplicate code that would otherwise result in your project quickly hitting the [script size limit](/workers/platform/limits/#worker-size). If you notice any bugs, let us know by filing a [GitHub issue](https://github.com/cloudflare/next-on-pages/issues/).
```

{{</Aside>}}

> ## Using a Static Export

Reviewer: ```suggestion
## Use a static export
```

> ### 1. Select your Next.js app

Reviewer: ```suggestion
### 1. Select your Next.js project
```

> If you already have a Next.js project that you wish to deploy, ensure that it is [configured for static exports](https://nextjs.org/docs/app/building-your-application/deploying/static-exports), change to its directory, and proceed to the next step. Otherwise, you can use `create-next-app` to start a new one.

Reviewer: ```suggestion
If you already have a Next.js project that you wish to deploy, ensure that it is [configured for static exports](https://nextjs.org/docs/app/building-your-application/deploying/static-exports), change to its directory, and proceed to the next step. Otherwise, use `create-next-app` to create a new Next.js project.
```

```sh
$ npx create-next-app --example with-static-export my-app
```

> After creating your project, a new `my-app` directory will be generated using the official [`with-static-export`](https://github.com/vercel/next.js/tree/canary/examples/with-static-export) example as a template. Switch to this directory to continue.

Reviewer: ```suggestion
After creating your project, a new `my-app` directory will be generated using the official [`with-static-export`](https://github.com/vercel/next.js/tree/canary/examples/with-static-export) example as a template. Change to this directory to continue.
```

   {{<table-wrap>}}

   | Configuration option | Value                              |
   | -------------------- | ---------------------------------- |
   | Production branch    | `main`                             |

>    | Build command        | `npx @cloudflare/next-on-pages@1` |

Reviewer 2: ```suggestion
   | Build command        | `npx @cloudflare/next-on-pages@1`  |
```

   | Build directory      | `.vercel/output/static`            |

   {{</table-wrap>}}

4. Next.js requires Node.js v16 or later to build successfully. To set your Node version, go to **Settings** in your Pages project > **Environment Variables (advanced)** section and add a `NODE_VERSION` variable with a value of `16` or greater.
5. Click on **Save and Deploy** to start the deployment. This first deployment will not be fully functional as the next step is also necessary.
6. In your Pages project, go to **Settings** > **Functions** > **Compatibility Flags**.

> 7. Configure a `nodejs_compat` flag for both production and preview .

Author: ```suggestion
7. Configure a `nodejs_compat` flag for both production and preview.
```

   {{<table-wrap>}}

   | Configuration option | Value           |
   | -------------------- | --------------- |
   | Production branch    | `main`          |
   | Build command        | `npm run build` |
   | Build directory      | `out`           |

   {{</table-wrap>}}

> 4. Next.js requires Node.js v16 or later to build successfully. To set your Node version, go to **Settings** in your Pages project > **Environment Variables (advanced)** section and add a `NODE_VERSION` variable with a value of `16` or greater.

Reviewer: ```suggestion
4. Next.js requires a specific Node.js version to build successfully. Refer to [System Requirements in Next.js Installation guide](https://nextjs.org/docs/getting-started/installation) to review the required Node.js version. To set your Node.js version, go to your Pages project > **Settings** > **Environment Variables (advanced)** section and add a `NODE_VERSION` variable with a value of the required version. For example, if the required Node.js version on Next.js's Installation guide is Node.js `16.8` or later, your environment variable value must be set to `16` or greater.
```

   {{<table-wrap>}}

   | Configuration option | Value           |
   | -------------------- | --------------- |
   | Production branch    | `main`          |
   | Build command        | `npm run build` |
   | Build directory      | `out`           |

   {{</table-wrap>}}

> 4. Next.js requires a specific Node.js version to build successfully. Refer to [System Requirements in Next.js Installation guide](https://nextjs.org/docs/getting-started/installation) to review the required Node.js version. To set your Node.js version, go to your Pages project > **Settings** > **Environment Variables (advanced)** section and add a `NODE_VERSION` variable with a value of the required version. For example, if the required Node.js version on Next.js's Installation guide is Node.js `16.8` or later, your environment variable value must be set to `16` or greater.

Reviewer 2: I am not completely sure, why not just telling people to set node to `16`? it is (currently) the correct value and we can always update this at any point (doing it in a generic way like this might be more future proof but to me it sounds like it could more of burden for people than a streigth "set node to 16")

Reviewer 2: Ah ok sorry this was suggested by @deadlypants1973, I still feel like it might add friction but I trust Kate's judgment so is she thinks this version's better I'm ok with it 🙂

Reviewer: We did this for another framework guide but basically, right now, we get PRs from external contributors that update this value. By telling the user to check the Installation guide, we are removing the need to update our docs with each Next.js update. The value of `16` is just an example but for the user to understand where to look, they should check [System Requirements in Next.js Installation guide](https://nextjs.org/docs/getting-started/installation). It's just a way for us to ease our maintenance requirements.

Reviewer 2: yeah I imagined that that was the reason, I'm fine with it, thanks for the clarification 🙂
